File: day-20/jurassic_jigsaw.py
import re, copy, itertools, random

input_file = 'day-20/input.txt'

directions = ['N','E','S','W']
deltas = dict(zip(['E','W','N','S'], [(1, 0), (-1, 0), (0, 1), (0, -1)]))

class Grid:
  def __init__(self, id, grid, flipped = False, rotated = 0):
    self.id = id
    self.flipped = flipped
    self.rotated = rotated
    self.grid = grid
    self.height = len(grid)
    self.width = len(grid[0])

  @classmethod
  def from_string(cls, input_string):
    lines = input_string.splitlines()
    id = re.match(r'Tile (\d+):', lines.pop(0))[1]
    grid = [[char for char in line] for line in lines]
    return cls(int(id), grid)

  def get_all_variants(self):
    variants = []
    variants.append(self)
    variants.append(variants[-1].rotate_90())
    variants.append(variants[-1].rotate_90())
    variants.append(variants[-1].rotate_90())
    variants.append(self.flip())
    variants.append(variants[-1].rotate_90())
    variants.append(variants[-1].rotate_90())
    variants.append(variants[-1].rotate_90())
    return variants

  def rotate_90(self):
    rotate = copy.deepcopy(self.grid)
    rotate = list(zip(*rotate[::-1]))  
    return Grid(self.id, rotate, self.flipped, self.rotated+1)

  def flip(self):
    return Grid(self.id, copy.deepcopy(self.grid[::-1]), not self.flipped, self.rotated)

  def get_matches(self, other):
    return [(alt, dir) for alt, dir in itertools.product(other.get_all_variants(), ['N','E','S','W']) if self.does_match(alt, dir)]

  def get_matches(self, other, dir):
    return [(alt, dir) for alt in other.get_all_variants() if self.does_match(alt, dir)]

  def does_match(self, other, direction):
    if direction == 'N':
      return ''.join(self.grid[0]) == ''.join(other.grid[-1])
    elif direction == 'E':
      return self.rotate_90().does_match(other.rotate_90(), 'S')
    elif direction == 'S':
      return ''.join(self.grid[-1]) == ''.join(other.grid[0])
    elif direction == 'W':
      return self.rotate_90().does_match(other.rotate_90(), 'N')

  def get_trim(self, border = 1):
    new_rows = []
    for i in range(border, len(self.grid)-border):
      new_rows.append(self.grid[i][border:len(self.grid[i])-border])
    return new_rows

  def count(self, char):
    count = 0
    for line in self.grid:
      count += ''.join(line).count(char)
    return count

  def __repr__(self):
    output = 'id: ' + str(self.id) + '\n'
    output += 'Flipped: ' + str(self.flipped) + '\n'
    output += 'Rotated: ' + str(self.rotated) + '\n'
    for y in range(self.height):
      for x in range(self.width):
        output += self.grid[y][x]
      output += '\n'
    return output

def calculate_layout(tiles):
  grid = {}
  x, y = 0, 0
  directions = ['N','E','S','W']

  grid[(x,y)] = tiles.pop(list(tiles.keys())[0])
  direction = directions[0]

  while direction != None and len(tiles) > 0: #All directions visited - done
    matches = []
    current_tile = grid[(x,y)]
    for tile in tiles.values():
      matches += current_tile.get_matches(tile, direction)
    if len(matches) == 0:
      new_x = x + deltas[dir][0]
      new_y = y + deltas[dir][1]
      if (new_x,new_y) in grid and grid[(new_x,new_y)] != None:
        x,y = new_x, new_y
        #must be back tracking?
      else:
        grid[(new_x,new_y)] = None
    elif len(matches) == 1:
      tile, dir = matches[0]
      x += deltas[direction][0]
      y += deltas[direction][1]
      grid[(x,y)] = tile
      tiles.pop(tile.id)
    else:
      print('mulitple matches!')
      return -10001    

    # Find a direction we haven't been in yet
    direction = None
    for dir in directions:
      new_x = x + deltas[dir][0]
      new_y = y + deltas[dir][1]
      if (new_x, new_y) not in grid:
        direction = dir
        break
    
    if direction == None: # All directions already explored - need to backtrack
      real_tiles = []
      for dir in directions:
        new_x = x + deltas[dir][0]
        new_y = y + deltas[dir][1]
        if grid[(new_x, new_y)] != None:
          real_tiles.append(dir)
      direction = random.choice(real_tiles)
      # Go to furtherest not None point in direction
      new_x = x + deltas[direction][0]
      new_y = y + deltas[direction][1]
      while (new_x, new_y) in grid and grid[(new_x, new_y)] != None:
        x = new_x
        y = new_y
        new_x = x + deltas[direction][0]
        new_y = y + deltas[direction][1]

  # trim Nones from grid
  grid = {key:val for key, val in grid.items() if val != None}
  return grid

def part1(input):
  tile_strings = input.split('\n\n')
  tiles = {}
  for string in tile_strings:
    tile = Grid.from_string(string)
    tiles[tile.id] = tile
  grid = calculate_layout(tiles)
   # find corners
  max_x = max(pos[0] for pos in grid.keys())
  min_x = min(pos[0] for pos in grid.keys())
  max_y = max(pos[1] for pos in grid.keys())
  min_y = min(pos[1] for pos in grid.keys())

  for (pos, tile) in grid.items():
    print('(' + ','.join([str(x) for x in pos]) + '):' + str(tile.id))
  return grid[(max_x, max_y)].id * grid[(min_x, max_y)].id * grid[(max_x, min_y)].id * grid[(min_x, min_y)].id

def part2(input):
  tile_strings = input.split('\n\n')
  tiles = {}
  for string in tile_strings:
    tile = Grid.from_string(string)
    tiles[tile.id] = tile
  grid = calculate_layout(tiles)

  max_x = max(pos[0] for pos in grid.keys())
  min_x = min(pos[0] for pos in grid.keys())
  max_y = max(pos[1] for pos in grid.keys())
  min_y = min(pos[1] for pos in grid.keys())

  tile_height = len(grid[(max_x, max_y)].grid)
  tile_width = len(grid[(max_x, max_y)].grid[0])
  
  print('\nlayout ids:')
  for y in range(max_y, min_y-1, -1):
    output = ''
    for x in range(min_x, max_x+1):
      output += str(grid[(x,y)].id)+ ' '
    print(output)

  print('\nlayout:')
  for y in range(max_y, min_y-1, -1):
    for r in range(0,tile_height): # Trims the row
      output = ''
      for x in range(min_x, max_x+1):
        output += ''.join(grid[(x,y)].grid[r])+ ' '
      print(output)
    print()

  print('\nstitched:')
  stitched = []
  for y in range(max_y, min_y-1, -1):
    for r in range(1,tile_height-1): # Trims the row
      row = []
      for x in range(min_x, max_x+1):
        row += grid[(x,y)].grid[r][1:tile_width-1]
      stitched.append(row)
  
  total = Grid(0, stitched)

  pattern = '''
                  # 
#    ##    ##    ###
 #  #  #  #  #  #   '''[1:].splitlines()

  patterns_found = max([count_patterns(alts.grid, pattern) for alts in total.get_all_variants()])

  return total.count('#') - (''.join(pattern).count('#') * patterns_found)

def count_patterns(grid, pattern):
  patterns = 0
  for row in range(len(grid) - len(pattern)):
    for col in range(len(grid[0]) - len(pattern[0])):
      found = True
      for y in range(len(pattern)):
        for x in range(len(pattern[0])):
          if pattern[y][x] == '#':
            found = found and grid[row+y][col+x] == '#'
      if found:
        patterns +=1
  return patterns

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-20/jurassic_jigsaw.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"
)

const inputFile string = "./day-20/input.txt"

type edge struct {
	normal, reverse, id string
}

func (e edge) flip() edge {
	return edge{normal: e.reverse, reverse: e.normal, id: e.id}
}

type edges struct {
	north, east, south, west edge
}

func reverse(input string) (result string) {
	for _, v := range input {
		result = string(v) + result
	}
	return
}

func createEdge(normal string) edge {
	reverse := reverse(normal)
	id := normal
	if reverse < id {
		id = reverse
	}
	return edge{normal: normal, reverse: reverse, id: id}
}

func getColumn(grid []string, col int) (column string) {
	for _, row := range grid {
		column += string(row[col])
	}
	return
}

func createEdges(grid []string) (edges edges) {
	edges.north = createEdge(grid[0])
	edges.east = createEdge(getColumn(grid, len(grid[0])-1))
	edges.south = createEdge(grid[len(grid)-1])
	edges.west = createEdge(getColumn(grid, 0))
	return
}

//Tile - struct to store info and functions of an image tile
type Tile struct {
	id      int
	grid    []string
	flipped bool
	rotated int
	edges   edges
}

//RotateClockwise - produces a new tile, which is the original rotated by 90 degrees clockwise
func (t Tile) RotateClockwise() Tile {
	rotated := []string{}
	for i := range t.grid {
		rowI := reverse(getColumn(t.grid, i))
		rotated = append(rotated, rowI)
	}
	rotatedEdges := edges{north: t.edges.west.flip(), east: t.edges.north, south: t.edges.east.flip(), west: t.edges.south}
	return Tile{id: t.id, flipped: t.flipped, grid: rotated, rotated: (t.rotated + 1) % 4, edges: rotatedEdges}
}

//FlipVertical - produces a new tile, which is the original mirrored along the vertical axis
func (t Tile) FlipVertical() Tile {
	flipped := []string{}
	for _, row := range t.grid {
		flipped = append(flipped, reverse(row))
	}
	flippedEdges := edges{north: t.edges.north.flip(), east: t.edges.west, south: t.edges.south.flip(), west: t.edges.east}
	return Tile{id: t.id, flipped: !t.flipped, grid: flipped, rotated: t.rotated, edges: flippedEdges}
}

// Rotate until we've done it 3 time, then reset and flip
func (t Tile) getTrimmed() []string {
	border := 1
	tileWidth := len(t.grid[0])
	trimmed := []string{}
	for i := border; i < len(t.grid)-border; i++ {
		trimmed = append(trimmed, t.grid[i][border:tileWidth-border])
	}
	return trimmed
}

func (t Tile) countStrings(search string) int {
	count := 0
	// searchRegex := regexp.MustCompile(search)
	for _, str := range t.grid {
		// matches := searchRegex.FindAllStringIndex(str, -1)
		// count += len(matches)
		count += strings.Count(str, search)
	}
	return count
}

// Rotate until we've done it 3 time, then reset and flip
func (t Tile) getNextArrangement() Tile {
	if t.rotated < 3 {
		return t.RotateClockwise()
	}
	return t.RotateClockwise().FlipVertical()
}

//CreateTile - created a tile object from a string
func CreateTile(tileString string) (int, Tile) {
	lines := strings.Split(strings.TrimSpace(tileString), "\n")
	var id int
	fmt.Sscanf(lines[0], "Tile %d:", &id)
	return id, Tile{id: id, flipped: false, grid: lines[1:], rotated: 0, edges: createEdges(lines[1:])}
}

func processInput(input string) map[int]Tile {
	tiles := make(map[int]Tile)
	tileStrings := strings.Split(strings.TrimSpace(input), "\n\n")
	for _, tileString := range tileStrings {
		id, tile := CreateTile(tileString)
		tiles[id] = tile
	}
	return tiles
}

func createEdgeMap(tiles map[int]Tile) map[string][]int {
	edgeMap := make(map[string][]int)
	for id, tile := range tiles {
		edges := tile.edges
		edgeMap[edges.north.id] = append(edgeMap[edges.north.id], id)
		edgeMap[edges.east.id] = append(edgeMap[edges.east.id], id)
		edgeMap[edges.south.id] = append(edgeMap[edges.south.id], id)
		edgeMap[edges.west.id] = append(edgeMap[edges.west.id], id)
	}
	return edgeMap
}

func identifyCornerIds(edgeMap map[string][]int) (corners []int) {
	openEdgeCount := map[int]int{}
	for _, tiles := range edgeMap {
		if len(tiles) == 1 {
			openEdgeCount[tiles[0]]++
		}
	}

	for id, count := range openEdgeCount {
		if count == 2 {
			corners = append(corners, id)
		}
	}
	return
}

func calculateLayout(tiles map[int]Tile) [][]Tile {
	edgeMap := createEdgeMap(tiles)

	gridLength := int(math.Ceil(math.Sqrt(float64(len(tiles)))))
	layout := [][]Tile{}

	// Pick top_left corner such that it has non-open edge to east and south
	topLeft := tiles[identifyCornerIds(edgeMap)[0]]
	for len(edgeMap[topLeft.edges.east.id]) == 1 || len(edgeMap[topLeft.edges.south.id]) == 1 {
		topLeft = topLeft.getNextArrangement()
	}

	for row := 0; row < gridLength; row++ {
		layout = append(layout, []Tile{})
		if row == 0 {
			layout[row] = append(layout[row], topLeft)
		} else {
			// Do north match to layout[row-1][0].south
			northTile := layout[row-1][0]
			northEdgeID, northID := northTile.edges.south.id, northTile.id

			var currentID int
			for _, currentID = range edgeMap[northEdgeID] {
				if currentID != northID {
					break
				}
			}
			currentTile := tiles[currentID]
			for currentTile.edges.north.normal != northTile.edges.south.normal {
				currentTile = currentTile.getNextArrangement()
			}

			layout[row] = append(layout[row], currentTile)
		}
		for col := 1; col < gridLength; col++ {
			// Do west match to layout[row][col-1].east
			westTile := layout[row][col-1]
			westEdgeID, westID := westTile.edges.east.id, westTile.id

			var currentID int
			for _, currentID = range edgeMap[westEdgeID] {
				if currentID != westID {
					break
				}
			}
			currentTile := tiles[currentID]
			for currentTile.edges.west.normal != westTile.edges.east.normal {
				currentTile = currentTile.getNextArrangement()
			}

			layout[row] = append(layout[row], currentTile)
		}
	}
	return layout
}

func stitchImage(layout [][]Tile) Tile {
	image := []string{}
	for _, row := range layout {
		rowOffset := len(image)
		for i, tile := range row {
			for j, gridLine := range tile.getTrimmed() {
				if i == 0 {
					image = append(image, "")
				}
				image[rowOffset+j] += gridLine
			}
		}
	}
	return Tile{id: 0, flipped: false, grid: image, rotated: 0, edges: createEdges(image)}
}

func countPatterns(image Tile, pattern []string) int {
	maxFound := 0
	for i := 0; i < 8; i++ {
		patterns := 0
		image = image.getNextArrangement()
		imageGrid := image.grid
		for row := 0; row < len(imageGrid)-len(pattern); row++ {
		search:
			for col := 0; col < len(imageGrid[0])-len(pattern[0]); col++ {
				for y := 0; y < len(pattern); y++ {
					for x := 0; x < len(pattern[0]); x++ {
						if pattern[y][x] == '#' {
							if imageGrid[row+y][col+x] != '#' {
								continue search
							}
						}
					}
				}
				patterns++
			}
		}
		if patterns > maxFound {
			maxFound = patterns
		}
	}
	return maxFound
}

//Part1 - Solution to Part 1 of the puzzle
func Part1(input string) int {
	tiles := processInput(input)
	edgeMap := createEdgeMap(tiles)

	product := 1
	for _, id := range identifyCornerIds(edgeMap) {
		product *= id
	}
	return product
}

//Part2 - Solution to Part 2 of the puzzle
func Part2(input string) int {
	tiles := processInput(input)
	layout := calculateLayout(tiles)
	image := stitchImage(layout)

	seaMonster := "..................#.\n#....##....##....###\n.#..#..#..#..#..#..."
	pattern := strings.Split(seaMonster, "\n")
	patternCount := countPatterns(image, pattern)

	return image.countStrings("#") - patternCount*strings.Count(seaMonster, "#")
}

func main() {
	b, err := ioutil.ReadFile(inputFile)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
